### PR TITLE
Raft improvements

### PR DESCRIPTION
**Pull request by dimmerex**
**Changes**
* Add raft to state so it does not disappear on game reload and re-launch
* Enable building raft to consume resources (2 wood and 1 rope)
* Enable player to dock and re-board raft
* Refactor

(Gif is sped up intentionally)
raft2

**Next steps**
* Add raft to task
* Update larger island (eg. add a coast too so the edge doesn't look jagged)

### CODE REVIEW

File: game/core/game_map.py [hunk 1]
```diff
@@ -24,14 +24,14 @@ def __init__(self, state, game_clock):  # type: ignore[no-untyped-def]
...
+    def draw(self) -> None:
         self.scene.draw()
         self.sparkles.draw()  # type: ignore[no-untyped-call]
 
-    def on_update(self):  # type: ignore[no-untyped-def]
+    def on_update(self) -> None:
         self.sparkles.on_update()
 
-    def load(self):  # type: ignore[no-untyped-def]
+    def load(self, reverse_movement=bool) -> None:
```
Comment (KyleKing) [inline game/core/game_map.py:34]:
```suggestion
    def load(self, reverse_movement: bool) -> None:
```

File: game/core/game_map.py [hunk 2]
```diff
@@ -67,14 +67,28 @@ def load(self):  # type: ignore[no-untyped-def]
 
         self.properties = my_map.properties
 
-        # Any layer with '_blocking' in it, will be a wall
         self.scene.add_sprite_list("wall_list", use_spatial_hash=True)
+
+        if not reverse_movement:
```
Comment (KyleKing) [inline game/core/game_map.py:72]: Maybe this could be named `inverese_movement`? Or `moving_on_water`?

File: game/core/game_map.py [hunk 3]
```diff
@@ -67,14 +67,28 @@ def load(self):  # type: ignore[no-untyped-def]
...
+        # Any layer not with 'water' or 'coast' will be a wall
         self.scene["wall_list"].clear()
         for layer, sprite_list in self.map_layers.items():
             if "water" not in layer and "coast" not in layer:
                 self.scene["wall_list"].extend(sprite_list)
+
+    def closest_land_coordinates(self, sprite):  # type: ignore[no-untyped-def]
+        # Get closest land coordinates when on water and trying to dock
+        closest = arcade.get_closest_sprite(sprite, self.scene["wall_list"])
+        return closest
```
Comment (KyleKing) [inline game/core/game_map.py:94]: Could be:

```suggestion
    def closest_land_coordinates(self, sprite: Sprite) -> Sprite:
        """Get closest land coordinates when on water and trying to dock."""
        return arcade.get_closest_sprite(sprite, self.scene["wall_list"])
```

File: game/core/game_state.py [hunk 4]
```diff
@@ -22,11 +36,14 @@ def __init__(self):  # type: ignore[no-untyped-def]
...
-        self.center_x = self.player_data["x"]
-        self.center_y = self.player_data["y"]
-        self.inventory = self.load_inventory(self.player_data["inventory"])
-        self.item = self.load_item(self.player_data["item"])
+        try:
+            self.player_data = self.get_player_data()
+            self.set_player_data()
+        except Exception as e:
+            logger.warning(e)
+            logger.warning(f"Failed to load player state, using default data...")
```
Comment (KyleKing) [inline game/core/game_state.py:44]: I would probably use `exception` if logging here:

```suggestion
        except Exception:
            logger.exception(f"Failed to load player state, using default data...")
```

File: game/core/game_state.py [hunk 5]
```diff
@@ -49,28 +66,61 @@ def get_layer_index(self, name) -> int | None:  # type: ignore[no-untyped-def]
...
+        self.center_x = self.player_data["x"]
+        self.center_y = self.player_data["y"]
+        self.inventory = self.load_inventory(self.player_data["inventory"])
+        self.item = self.load_item(self.player_data["item"])
+        self.vehicle = None
+        if self.player_data["vehicle_type"]:
+            self.vehicle = self.load_vehicle(self.player_data)
+            self.vehicle_x = self.player_data["vehicle_x"]
+            self.vehicle_y = self.player_data["vehicle_y"]
+            self.vehicle_docked = self.player_data["vehicle_docked"]
```
Comment (KyleKing) [inline game/core/game_state.py:80]: We use the term `vehicle`, but this only ever refers to a boat. Maybe we could do a find+replace with `boat`, `watercraft`, or something else to indicate that it is something that can only move on water

File: game/core/game_state.py [hunk 6]
```diff
@@ -49,28 +66,61 @@ def get_layer_index(self, name) -> int | None:  # type: ignore[no-untyped-def]
...
+            return {
+                "x": STARTING_X,
+                "y": STARTING_Y,
+                "inventory": [],
+                "item": None,
+                "vehicle_type": None,
+                "vehicle_x": None,
+                "vehicle_y": None,
+                "vehicle_docked": False,
+            }
```
Comment (KyleKing) [inline game/core/game_state.py:101]: Can this be moved to the `constants` file as the `DEFAULT_PLAYER_DATA`?

File: game/core/models/base_player_inventory.py [hunk 7]
```diff
@@ -88,6 +88,11 @@ def equip_item(self, item_name: str) -> Sprite:
         self.equipped_item = self.inventory[item_name]
         return self.equipped_item.sprite
 
+    @beartype
+    def unequip_item(self) -> None:
+        """Unequip current item."""
+        self.equipped_item = None
```
Comment (KyleKing) [inline game/core/models/base_player_inventory.py:94]: We should probably use `store_equipped_item` instead

https://github.com/DesignPatternsAdventure/game/blob/ddd7f1ba38fd029a335146ed2269c20bc15321fb/game/core/models/base_player_inventory.py#L79-L83

**Review by KyleKing — APPROVED** [review]
This is awesome!

File: game/core/game_map.py (hunk 2 above)
Comment (dimmerex) [reply game/core/game_map.py:72]: Good idea, I like `inverse_movement`!

File: game/core/game_state.py (hunk 5 above)
Comment (dimmerex) [reply game/core/game_state.py:68]: I was thinking the game itself could potentially have land vehicles in the future, although the implementation would need to be adjusted (eg. `vehicle_docked` would be part of some `vehicle_metadata`

File: game/core/models/base_player_inventory.py (hunk 7 above)
Comment (dimmerex) [reply game/core/models/base_player_inventory.py]: Ah okay! I deleted the new method and renamed `store_equipped_item` to `unequip_item` to be more straightforward. Let me know if there are any concerns, I can revert it back easily.

**Review by KyleKing — APPROVED** [review]
LGTM 👍🏻

**Pull request merged**